p2p/src/peer.rs

```diff
@@ -38,328 +41,377 @@ const MAX_HANDSHAKE_LENGTH: usize = 255;
+        Connection {
+            read: Some(read),
+            write: Some(write),
+            ..Connection::default()
+        }
+    }
+}
+
+/// Peer's cryptographic data
+pub struct Cryptographer<T, K, E>
 where
-    T: Debug + Encode + Decode + Send + Clone + 'static,
+    T: Encode + Decode + Send + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
-    /// Peer identification
-    pub id: PeerId,
-    /// A unique connection id
-    pub connection_id: ConnectionId,
-    /// Reading half of `TcpStream`
-    pub read: Option<OwnedReadHalf>,
-    /// Writing half of `TcpStream`
-    pub write: Option<OwnedWriteHalf>,
-    /// Current peer/connection state
-    pub state: State,
-    /// Flag stating that this connection is outgoing
-    pub outgoing: bool,
     /// Secret part of keypair
-    pub secret_key: PrivateKey,
+    secret_key: PrivateKey,
     /// Public part of keypair
-    pub public_key: PublicKey,
+    public_key: PublicKey,
     /// Encryptor created from session key, that we got by Diffie-Hellman scheme
-    pub cipher: Option<SymmetricEncryptor<E>>,
-    /// The to send received messages upstairs
-    pub broker: Broker,
-    /// A flag that stops listening stream
-    finish_sender: Option<Sender<()>>,
+    cipher: Option<SymmetricEncryptor<E>>,
     /// Phantom
-    pub _key_exchange: PhantomData<K>,
+    _key_exchange: PhantomData<K>,
     /// Phantom2
-    pub _post_type: PhantomData<T>,
+    _post_type: PhantomData<T>,
 }
 
-impl<T, K, E> Peer<T, K, E>
+impl<T, K, E> Cryptographer<T, K, E>
 where
-    T: Debug + Encode + Decode + Send + Clone + 'static,
+    T: Encode + Decode + Send + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
-    /// Creates a peer
-    /// # Errors
-    /// If it can not create a keypair, so never
-    pub fn new_inner(
-        id: PeerId,
-        connection_id: ConnectionId,
-        stream: Option<TcpStream>,
-        state: State,
-        broker: Broker,
-    ) -> Result<Self, Error> {
+    fn default_or_err() -> Result<Self, Error> {
         // P2P encryption primitives
         let dh = K::new();
-        let (public_key, secret_key) = match dh.keypair(None) {
-            Ok((public_key, secret_key)) => (public_key, secret_key),
-            Err(e) => {
-                warn!(%e, "Error generating keypair");
-                return Err(Error::Keys);
-            }
-        };
-        // If we are connected we take apart stream for two halves.
-        // If we are not connected we save Nones and wait for message to start connecting.
-        let (read, write) = match stream.map(TcpStream::into_split) {
-            None => (None, None),
-            Some((read, write)) => (Some(read), Some(write)),
-        };
-        let outgoing = read.is_none() && write.is_none();
+        let (public_key, secret_key) = dh.keypair(None).log_warn("Error generating keypair")?;
```

Comment:
Like here for example. Why log it if we still return it? We will probably `log` or `expect` at the place where we call this.

Reply:
In the `p2p` crate the `?` operator is used mostly for short-circuiting. The logging is used to notify if the error took place and was immensely helpful in debugging, especially since it can also produce the line number of the place where the error took place.